share/spack/csh/spack.csh

```diff
@@ -78,51 +78,67 @@ case env:
     set _sp_arg=""
     [ $#_sp_args -gt 0 ] && set _sp_arg = ($_sp_args[1])
 
-    if ( "$_sp_arg" == "-h" ) then
+    if ( "$_sp_arg" == "-h" || "$_sp_arg" == "--help" ) then
         \spack env -h
     else
         switch ($_sp_arg)
             case activate:
                 set _sp_env_arg=""
                 [ $#_sp_args -gt 1 ] && set _sp_env_arg = ($_sp_args[2])
 
-                if ( "$_sp_env_arg" == "" || "$_sp_args" =~ "*--sh*" || "$_sp_args" =~ "*--csh*" || "$_sp_args" =~ "*-h*" ) then
-                    # no args or args contain -h/--help, --sh, or --csh: just execute
+                # Space needed here to differentiate between `-h`
+                # argument and environments with "-h" in the name.
+                if ( "$_sp_env_arg" == "" || \
+                     "$_sp_args" =~ "* --sh*" || \
+                     "$_sp_args" =~ "* --csh*" || \
+                     "$_sp_args" =~ "* -h*" || \
+                     "$_sp_args" =~ "* --help*" ) then
+                    # No args or args contain --sh, --csh, or -h/--help: just execute.
                     \spack $_sp_flags env $_sp_args
                 else
                     shift _sp_args  # consume 'activate' or 'deactivate'
-                    # actual call to activate: source the output
+                    # Actual call to activate: source the output.
                     eval `\spack $_sp_flags env activate --csh $_sp_args`
                 endif
                 breaksw
             case deactivate:
                 set _sp_env_arg=""
                 [ $#_sp_args -gt 1 ] && set _sp_env_arg = ($_sp_args[2])
 
-                if ( "$_sp_env_arg" != "" ) then
-                    # with args: execute the command
+                # Space needed here to differentiate between `--sh`
+                # argument and environments with "--sh" in the name.
+                if ( "$_sp_args" =~ "* --sh*" || \
+                     "$_sp_args" =~ "* --csh*" ) then
+                    # Args contain --sh or --csh: just execute.
                     \spack $_sp_flags env $_sp_args
+                else if ( "$_sp_env_arg" != "" ) then
+                    # Any other arguments are an error or -h/--help: just run help.
+                    \spack $_sp_flags env deactivate -h
                 else
-                    # no args: source the output
+                    # No args: source the output of the command.
                     eval `\spack $_sp_flags env deactivate --csh`
                 endif
                 breaksw
             default:
-                echo default
                 \spack $_sp_flags env $_sp_args
                 breaksw
         endsw
     endif
+    breaksw
```

Comment:
This single missing line turns out to be extremely important. Without this, the current behavior on develop looks like:
```console
% spack env -h
usage: spack env [-h] SUBCOMMAND ...

manage virtual environments

positional arguments:
  SUBCOMMAND
    activate   set the current environment
    deactivate
               deactivate any active environment in the shell
    create     create a new environment
    remove (rm)
               remove an existing environment
    list (ls)  list available environments
    status (st)
               print whether there is an active environment
    loads      list modules for an installed environment '(see spack module loads)'
    view       manage a view associated with the environment

optional arguments:
  -h, --help   show this help message and exit
usage: spack env [-h] SUBCOMMAND ...

manage virtual environments

positional arguments:
  SUBCOMMAND
    activate   set the current environment
    deactivate
               deactivate any active environment in the shell
    create     create a new environment
    remove (rm)
               remove an existing environment
    list (ls)  list available environments
    status (st)
               print whether there is an active environment
    loads      list modules for an installed environment '(see spack module loads)'
    view       manage a view associated with the environment

optional arguments:
  -h, --help   show this help message and exit
% spack env list
default
==> 4 environments
    test--sh  test-h  test-sh  test-z
==> Error: unrecognized arguments: --csh
```
All commands are run twice, because it hits the default below.